src/uu/tail/src/tail.rs

```diff
@@ -125,159 +200,346 @@ impl Settings {
+                unbounded_tail(&mut reader, &settings)?;
+                if settings.follow == Some(FollowMode::Descriptor) {
+                    // Insert `stdin` into `files.map`
+                    files.insert(
+                        path.to_path_buf(),
+                        PathData {
+                            reader: Some(Box::new(reader)),
+                            metadata: None,
+                            display_name: PathBuf::from(text::STDIN_HEADER),
+                        },
+                    );
                 }
             } else {
-                let mut reader = BufReader::new(file);
-                unbounded_tail(&mut reader, settings)?;
-                if settings.follow {
-                    readers.push((Box::new(reader), filename));
+                set_exit_code(1);
+                show_error!(
+                    "cannot fstat {}: {}",
+                    text::STDIN_HEADER.quote(),
+                    text::BAD_FD
+                );
+                if settings.follow.is_some() {
+                    show_error!(
+                        "error reading {}: {}",
+                        text::STDIN_HEADER.quote(),
+                        text::BAD_FD
+                    );
                 }
             }
+        } else if path_is_tailable {
+            match File::open(&path) {
+                Ok(mut file) => {
+                    if settings.verbose {
+                        files.print_header(&path, !first_header);
+                        first_header = false;
+                    }
+                    let mut reader;
+
+                    if file.is_seekable() && metadata.as_ref().unwrap().get_block_size() > 0 {
+                        bounded_tail(&mut file, &settings);
+                        reader = BufReader::new(file);
+                    } else {
+                        reader = BufReader::new(file);
+                        unbounded_tail(&mut reader, &settings)?;
+                    }
+                    if settings.follow.is_some() {
+                        // Insert existing/file `path` into `files.map`
+                        files.insert(
+                            path.canonicalize()?,
+                            PathData {
+                                reader: Some(Box::new(reader)),
+                                metadata,
+                                display_name,
+                            },
+                        );
+                    }
+                }
+                Err(e) if e.kind() == std::io::ErrorKind::PermissionDenied => {
+                    show!(e.map_err_context(|| {
+                        format!("cannot open {} for reading", display_name.quote())
+                    }));
+                }
+                Err(e) => {
+                    return Err(e.map_err_context(|| {
+                        format!("cannot open {} for reading", display_name.quote())
+                    }));
+                }
```

Comment:
Why do we want to only print the error in one case and return in the other?

Reply:
I answered this [here](https://github.com/uutils/coreutils/pull/2695#discussion_r890025691). Basically, if we open multiple files and one has PermissionDenied, just print the error and continue with the other files. This is also what GNU's tail does.
```bash
$ tail file_a missing file_b unreadable file_c
==> file_a <==
A
tail: cannot open 'missing' for reading: No such file or directory

==> file_b <==
B
tail: cannot open 'unreadable' for reading: Permission denied

==> file_c <==
C
```
I haven't checked the behaviour of all possible error types, just the ones that came up in GNU's test-suite.

Reply:
My bad for forgetting you already answered that! Great work!